A74-xds-config-tears.md

```diff
@@ -0,0 +1,349 @@
+  virtual void OnError(absl::string_view context, absl::Status status) = 0;
+  virtual void OnResourceDoesNotExist(std::string context) = 0;
+};
+```
+
+The xds resolver will implement this watcher interface and pass the
+watcher object to the XdsDependencyManager at instantiation time.
+
+The updates delivered to this watcher will contain a complete xDS
+configuration, including all necessary resources.  The xDS configuration
+will look something like this (C++ syntax, various implementation
+details elided):
+
+```c++
+struct XdsConfig {
+  // Listener resource.
+  XdsListenerResource listener;
+  // RouteConfig resource.  Will be populated even if RouteConfig is
+  // inlined into the Listener resource.
+  XdsRouteConfigResource route_config;
+  // Virtual host.  Points into route_config.  Will always be non-null.
+  dsRouteConfigResource::VirtualHost* virtual_host;
+
+  // Cluster map.  A cluster will have a non-OK status if either
+  // (a) there was an error and we did not already have a valid
+  // resource or (b) the resource does not exist.
+  struct ClusterConfig {
+    // Cluster name and resource.
+    std::string cluster_name;
+    XdsClusterResource cluster;
+
+    // Endpoint info.  If there was an error, endpoints will be null
+    // and resolution_note will be set.
+    struct EndpointConfig {
+      XdsEndpointResource endpoints;
+      std::string resolution_note;
+    };
+    // The list of leaf clusters for an aggregate cluster.
+    struct AggregateConfig {
+      std::vector<absl::string_view> leaf_clusters;
+    };
+    std::variant<EndpointConfig, AggregateConfig> children;
+  };
+  std::map<std::string, absl::StatusOr<ClusterConfig>> clusters;
+};
+```
+
+Note that the watcher's OnError() and OnResourceDoesNotExist()
+methods will be invoked only for problems with the listener and route
+configuration resources.  For cluster-specific errors, we need to
+retain the existing behavior, which is that it causes a problem for the
+individual cluster but does not affect traffic sent to any other cluster.
+Therefore, when the XdsDependencyManager sees an error on a cluster for
+which it did not previously have data, or if it sees a does-not-exist
+notification on a cluster, it will report that by returning a non-OK
+status for the cluster in the XdsConfig.  Similarly, if there is a problem
+obtaining endpoint data from EDS or DNS, the XdsDependencyManager will
+report that problem via the cluster's resolution_note field.
+
+The xds resolver will pass the returned XdsConfig to the LB policies via
+an attribute, so that the LB policies can use the cluster data.  Note that
+we cannot simply pass the data via the cds LB policy's JSON config,
+because we need this to work in cases where the cluster is determined
+dynamically, such as where a route uses a ClusterSpecifierPlugin, in
+which case the LB policy's config comes from the ClusterSpecifierPlugin,
+not from the xds resolver.  (We have some internal functionality that
```

Comment:
I'm not following this fully.

ClusterSpecifierPlugin config is already determined in the xds resolver, is it not?  That creates a child cluster with the name `cluster_specifier_plugin:<plugin_name>` and using the child policy config generated by the CSP.  We do this here in Go:

https://github.com/grpc/grpc-go/blob/e20d5ce8c48ee70cfabf6d7dc73c4367ee529e04/xds/internal/resolver/xds_resolver.go#L312-L319

It sounds like you're saying some part of this is per-RPC and determines the structure of the LB policy tree somehow?

Reply:
When a route uses a ClusterSpecifierPlugin, the xds resolver does indeed create a `cluster_specifier_plugin:<plugin_name>` entry in the LB policy tree, which configures the policy for that plugin as a child of the xds_cluster_manager policy.  However, that plugin policy will then dynamically create cds policies as its children on the basis on individual RPCs.  The LB policy config for those cds child policies is generated by the ClusterSpecifierPlugin policy, not by the xds resolver, which is why we need some other way to pass the configuration down to them.